feat: Support for Ubuntu Pro/Landscape configurations

ashuntu opened the pull request:
This is a draft PR against this fork for the sake of iteratively working on changes.

ashuntu committed:
Fix dscheck for WSL

Previously dscheck would fail due to the for loop not properly
expanding. This change removes the quotes around $@ and corrects
some spelling mistakes to fix that.

ashuntu committed:
feat(WSL): Add support for Landscape and agent configs

ashuntu committed:
Fix ds-identify WSL tests

ds-identify changes in the previous commits broke its tests, so
this corrects those tests and accounts for the new ds-identify
which considers the new directory locations. This also breaks
up ds-identify into more functions to allow for better testing.

Co-authored-by: Carlos Nihelton <carlosnsoliveira@gmail.com>

ashuntu committed:
Parse and load non-YAML config files

Previously, configuration files were assumed to be YAML only. This
change allows for non-YAML files to be loaded by just setting
userdata_raw to a list of binary data instead of attempting to parse
each config manually. This has the consequence of not being able
to control how merging works for those files.

cnihelton commented on tools/ds-identify:
```diff
@@ -1594,21 +1594,42 @@ dscheck_VMware() {
-            if [ -n "$cloudinitdir" ]; then
+            WSL_run_cmd "echo %USERPROFILE%"
+            profiledir="{$_RET}"
+            profiledir="${profiledir%%[[:cntrl:]]}"
```
```suggestion
            profiledir="${_RET%%[[:cntrl:]]}"
```

We can make this a single line and also the syntax in the first line is incorrect: `${_RET}`.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -128,6 +129,28 @@ def cloud_init_data_dir() -> PurePath:
+    if not home:
+        raise subp.ProcessExecutionError(
+            "No output from cmd.exe to show the user profile dir."
+        )
```
Finding the  `%USERPROFILE%` directory can be somewhat expensive, so I think we should compute it early in the datasource `_get_data()` method and pass it as argument to those functions that are repeating this algorithm (`ubuntu_pro_data_dir` and `cloud_init_data_dir` ).

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -179,6 +216,59 @@ def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
+def load_landscape_data(instance_name: str) -> dict | bytes | None:
+    """
+    Load Landscape config data into a dict, returning an empty dict if nothing
+    is found. If the file is not a YAML, returns the binary file.
```
```suggestion
    is found. If the file is not a YAML, returns the raw binary file contents.
```

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -179,6 +216,59 @@ def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
+
+def load_agent_data() -> dict | bytes:
+    """
+    Load agent.yaml data into a dict, returning an empty dict if nothing is found. If the file is not a YAML, returns the binary file.
```
I'd expect the linter to warn you of this being a too long line (over 79 chars).

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -179,6 +216,59 @@ def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
+
+        return agent_data
+    except FileNotFoundError:
+        LOG.debug("No Pro agent data found at %s, ignoring.", data_path)
```
We can rearrenge the `load_landscape_data` and `load_agent_data` functions so that this algorithm is not repeated, but rather, called by them with the relevant file paths.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -239,20 +329,45 @@ def check_instance_id(self, sys_cfg) -> bool:
-            LOG.error("Unable to setup WSL datasource: %s", str(err))
-            return False
+            LOG.error("Unable to load user data: %s", str(err))
+
```
Although ds-identify prevents that, we must still work with the possibility of missing both agent and user data files, thus setting `self.userdata_raw` to `None` and returning `False`. The current implementation does not allow to correctly distinguish between empty files and missing files.

So, similar to the `load_landscape_data` function we should return None from `load_agent_data` as well as `load_user_data` meaning that the relevant file does not exist, and in case all data loaded is None we can bail out.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -239,20 +329,45 @@ def check_instance_id(self, sys_cfg) -> bool:
-            return False
+            LOG.error("Unable to load user data: %s", str(err))
+
+        should_list = isinstance(agent_data, bytes) or isinstance(user_data, bytes)
+        if should_list:
+            self.userdata_raw = [user_data, agent_data]
+            return True
```
This deserves a comment: if the loaded data cannot be modelled as dicts, we can just pass a List containing the parts and cloud-init internally handles merging them by building some sort of MIMEMultipart object.

cnihelton commented on tests/unittests/test_ds_identify.py:
```diff
@@ -1183,42 +1183,84 @@ def test_no_fs_mounts(self):
         return self._check_via_dict(data, RC_NOT_FOUND)
 
+    def test_found_fail_due_instance_name_parsing(self):
+        """WLS datasource detection fail due parsing error even though the file exists."""
```
```suggestion
        """WSL datasource detection fail due parsing error even though the file exists."""
```

cnihelton commented on tests/unittests/test_ds_identify.py:
```diff
@@ -1183,42 +1183,84 @@ def test_no_fs_mounts(self):
+        Path(filename).unlink()
+
     def test_found_via_userdata_version_codename(self):
         """WLS datasource detected by VERSION_CODENAME when no VERSION_ID"""
```
```suggestion
        """WSL datasource detected by VERSION_CODENAME when no VERSION_ID"""
```

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -179,6 +216,59 @@ def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
+    return {}
+
+
+def load_user_data() -> dict | bytes | None:
+
+
```
Just so we don't loose track: implementation is missing.

cnihelton commented on tests/unittests/test_ds_identify.py:
```diff
@@ -1229,15 +1271,25 @@ def test_found_via_userdata(self):
+                "RET": userprofile,
             },
         )
+        cloudinitdir = os.path.join(userprofile, ".cloud-init")
+        os.mkdir(cloudinitdir)
+        up4wcloudinitdir = os.path.join(userprofile, ".ubuntupro/.cloud-init")
+        os.makedirs(up4wcloudinitdir, exist_ok=True)
         userdata_files = [
+            os.path.join(
+                up4wcloudinitdir, MOCK_WSL_INSTANCE_DATA["name"] + ".user-data"
+            ),
+             os.path.join(
+                up4wcloudinitdir, "agent.yaml"
+            ),
```
We need to add something similar to this in the datasource unit tests, so we can assert that the pro-provided user data files are loaded with the correct precedence.

ashuntu replied:
I added the beginnings of a test, though I realize it isn't complete.

A lot of the tests are currently failing due to unexpected calls to subp, so I'm assuming we need to play around with some mocks

ashuntu committed:
Simplify profiledir variable

ashuntu committed:
Reduce amount of cmd.exe calls

ashuntu committed:
Fix comments and formatting

ashuntu committed:
Add myself to cla-signers

ashuntu committed:
Add file loading utility function

Breaks up yaml/binary loading into a function for better
reusability. Also adds more context to the usage of a list
instead of manually merging dicts from YAML.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -119,15 +120,35 @@ def cloud_init_data_dir() -> PurePath:
             "No output from cmd.exe to show the user profile dir."
         )
+    return home
+
```
Why not return the path already translated to Linux? Notice that both `cloud_init_data_dir` and `ubuntu_pro_data_dir` are required to perform the translation.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -119,15 +120,35 @@ def cloud_init_data_dir() -> PurePath:
-    win_profile_dir = win_path_2_wsl(home)
+def cloud_init_data_dir(user_home: str) -> PurePath:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
```
Outdated docstring.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -179,6 +232,42 @@ def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
     return metadata
 
 
+def load_landscape_data(
+    instance_name: str, user_home: str
+) -> dict | bytes | None:
+    """
+    Load Landscape config data into a dict, returning an empty dict if nothing
+    is found. If the file is not a YAML, returns the raw binary file contents.
+    """
+    data_dir = ubuntu_pro_data_dir(user_home)
+    if data_dir is None:
+        return {}
+
+    data_path = os.path.join(
+        data_dir.as_posix(), landscape_file_name(instance_name)
+    )
+
+    return load_yaml_or_bin(data_path)
+
+
+def load_agent_data(user_home: str) -> dict | bytes | None:
+    """
+    Load agent.yaml data into a dict, returning an empty dict if nothing is
+    found. If the file is not a YAML, returns the raw binary file contents.
+    """
+    data_dir = ubuntu_pro_data_dir(user_home)
+    if data_dir is None:
+        return {}
+
+    data_path = os.path.join(data_dir.as_posix(), agent_file_name())
+
+    return load_yaml_or_bin(data_path)
```
In both cases if the `ubuntu_pro_data_dir` is None the files won't exist, thus we should return None, as that's how the check in `_get_data` understands it.

cnihelton reviewed (changes requested):
We're getting there, but there is still some work left.

ashuntu committed:
Add WSL datasource unit test

ashuntu committed:
Properly handle cloud-init files not found

Previously if a ~/.cloud-init dir was not found then we'd improperly
error out. Since we've added Ubuntu Pro configs, that's not always
a bad thing so this change allows for no ~/.cloud-init to be found
and userdata will still attempt to load from other locations.

ashuntu committed:
Fix lint problems and Ruff formatting

Reorder test imports

Fix variable types and formatting

ashuntu committed:
Fix pecedence test and dir mocks

The main precedence test and others unittests were broken due to
previous changes. This change fixes those tests somewhat, but a
commit after this will permanently fix the problems.

ashuntu commented:
I think the only thing left for tests is the mimetype not being correct, maybe that's something we can address quickly together @CarlosNihelton. I could also create a few difference cases for precedence testing, but right now the test with every cloud-init file available works well.

ashuntu committed:
Disable shellcheck for profile directory expansion

Since we do actually need $@ to expand to properly handle the
directories passed in by the caller, we can safetly disable
shellcheck here.

Co-authored-by: Carlos Nihelton <carlosnsoliveira@gmail.com>

ashuntu committed:
Increment precedence testing

Adds more incremental tests to userdata loading to account for different
machine configurations, including the new cloud-init configs provided
by Ubuntu Pro.

ashuntu committed:
Satisfy mypy and black

Minor refactor and renaming

ashuntu commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -238,22 +327,66 @@ def check_instance_id(self, sys_cfg) -> bool:
-        seed_dir = cloud_init_data_dir()
+        user_home = find_home()
+        seed_dir = cloud_init_data_dir(user_home)
+        user_data: Optional[Union[dict, bytes]] = {}
```
Should this be `= None` instead of an empty dict? I think making it `None` only helps if `load_landscape_data` fails for some reason (which should essentially never occur), in which case `userdata_raw` can be `None` and we early return.

cnihelton replied:
I think making it None makes more sense because we avoid the "cost" of creating a dict for no reason.

ashuntu commented:
Some checks are still failing, but I think they're unrelated? Otherwise, this should be good to go unless there is any more feedback.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -238,22 +327,66 @@ def check_instance_id(self, sys_cfg) -> bool:
+            user_data = load_landscape_data(self.instance_name, user_home)
+            if user_data is None:
+                # Regular user data
+                if seed_dir is None:
+                    raise ValueError
+                file = self.find_user_data_file(seed_dir)
+                if os.path.exists(file.as_posix()):
+                    bin_user_data = util.load_binary_file(file.as_posix())
+                    user_data = util.load_yaml(bin_user_data)
+                    user_data = (
+                        bin_user_data if user_data is None else user_data
+                    )
```
Instead of raising the exception here, I'd rather nest the code below under the if-clause. Raising an exception just for skipping a piece of code is a mistake. The datasource should be OK with finding just the agent data with no user data at all, that's not an exceptional situation.

```suggestion
                if seed_dir is not None:
                    file = self.find_user_data_file(seed_dir).as_posix()
                    if os.path.exists(file):
                        bin_user_data = util.load_binary_file(file)
                        user_data = util.load_yaml(bin_user_data)
                        user_data = (
                            bin_user_data if user_data is None else user_data
                        )
```

If you think this is too nested, then we should extract part of it inside a new method.

cnihelton commented on cloudinit/sources/DataSourceWSL.py:
```diff
@@ -238,22 +327,66 @@ def check_instance_id(self, sys_cfg) -> bool:
+        user_home = find_home()
+        seed_dir = cloud_init_data_dir(user_home)
+        user_data: Optional[Union[dict, bytes]] = {}
+        should_list = False
```
Better name for this variable: `requires_multipart`.

cnihelton reviewed (changes requested):
We are almost good to go: please address the last two comments, rebase over the last main organizing the commits to tell your story more easily to the reviewers and let's open the PR against upstream :tada:

ashuntu committed:
Update WSL datasource documentation

ashuntu committed:
Fix WSL_run_cmd relying on external variables

ashuntu committed:
Add clarifying comment regarding variable expansion

ashuntu committed:
Simplify loading of Ubuntu Pro data files

Created a unified load_ubuntu_pro_data to share some of the logic. Also changed functions for getting file names into const variables.

_get_data separates some of the logic out of the single try/except to prevent early termination of certain code paths. load_ubuntu_pro_data is robust enough to only fail when it should.

ashuntu committed:
Remove unnecessary wrapper for Ubuntu Pro dir

ashuntu committed:
Merge wslpath functions into a single function

ashuntu committed:
Correct grammar in WSL docs

ashuntu committed:
Inline wsl_path_2_win

ashuntu committed:
Inline win_path_2_wsl and add exceptions to doc comment

ashuntu committed:
Instantiate instance name in _get_data

ashuntu committed:
Add additional environment checks before attempting to get data

ashuntu committed:
Streamline raw user_data cases

ashuntu committed:
Log merged keys, instead of full merged data

ashuntu committed:
Check for Ubuntu distro before looking for Pro data

ashuntu committed:
Correct linting and utilize datasource distro

blackboxsw committed:
tests: adding missed mocks subp.which, refactor using common fixture

- add missing subp.which mock of WSLPATH_CMD
- provide non-empty distro to all DataSourceWSL init
- consolidate common mocks in autouse pytest fixture
